### DISCUSSION TITLE
High availability deploy

### DISCUSSION THREAD

**Discussion by AlliotTech** [danny-avila/LibreChat#5099] [Q&A] [answered]
Hi, it seems like the LibreChat API container doesn't support HA deployment (multiple replicas).right?

**Comment by AlliotTech** [comment]
I enabled AzureAD login.

When I deployed 2 replicas for librechat-api, I sometimes get the following error when trying to log in with Azure AD: 
```
Error: did not find expected authorization request details in session, req.session["oidc:login.microsoftonline.com"] is undefined
```

**Comment by AlliotTech** [comment]
```
Warning: connect.session() MemoryStore is not
designed for a production environment, as it will leak
memory, and will not scale past a single process.
```

I believe the root cause is the memoryStore session used by the librechat container.

**Reply by AlliotTech** [reply]
ref: https://github.com/danny-avila/LibreChat/discussions/1840

**Reply by danny-avila** [reply]
You need to use Redis if using replicas

**Reply by AlliotTech** [reply]
Is there any documentation about this feature? The official documentation only states that it is not production-ready, but high availability is very important.

**Reply by danny-avila** [reply, marked as answer]
https://www.librechat.ai/docs/configuration/dotenv#redis

**Reply by AlliotTech** [reply]
Thanks for your help!   For the HA deployment, the LibreChat API container relies on local storage for image storage. Are there any plans to support S3 or other distributed storage?

**Reply by danny-avila** [reply]
Yes. Supporting S3 will cover a lot of different storage options for S3-compatible services. I’m also hoping to implement MCP resources which will add plug-and-play storage options
